internal/auth/password/argon2.go

```diff
@@ -7,16 +7,46 @@ import (
 	"context"
 	"crypto/rand"
 	"strings"
+	"sync"
 
 	"github.com/hashicorp/boundary/internal/auth/password/store"
 	"github.com/hashicorp/boundary/internal/errors"
 	"github.com/hashicorp/boundary/internal/oplog"
 	wrapping "github.com/hashicorp/go-kms-wrapping/v2"
 	"github.com/hashicorp/go-kms-wrapping/v2/extras/structwrapping"
+	"github.com/hashicorp/go-secure-stdlib/permitpool"
 	"golang.org/x/crypto/argon2"
 	"google.golang.org/protobuf/proto"
 )
 
+var (
+	// hashingPermitPool is the global permit pool used to restrict concurrent
+	// password hashing. It can be configured with SetHashingPermits.
+	hashingPermitPool *permitpool.Pool
+	// hashingPermitPoolLock is used to synchronize access to hashingPermitPool
+	// This is an RWMutex to allow an unlimited number of readers
+	// and a single writer, since allowing a single reader or writer
+	// would effectively make the pool useless.
+	hashingPermitPoolLock *sync.RWMutex
```

Comment:
If I am following this correctly, this mutex is use to allow the entire permit pool to be replaced safely, allowing it to be resized. Could this mutex be internalized into the pool itself by exposing a Resize method on the pool?

Reply:
I think I'd prefer not to add this to the pool itself, as it may encourage the use of global pools like this one. Ideally a global pool like this isn't necessary, and all operations can be funneled through a single type or struct, which would then allow each instance of such a struct to have a pool assigned on it (and the size decided at runtime). The only reason we need to resize this pool is because it is a global variable with a default size set at init time. Does that makes sense?

Reply:
Even if it isn't a global variable, wouldn't any resize operation require a mutex? And there could be other times that a resize is necessary or desirable, like if the pool size can be changed via a SIGHUP.

Reply:
Yes, resizing in general would require a mutex, I agree. But from the permitpool package point of view, a pool can just be considered immutable, and if a user wants to resize it they have to replace it. That does punt some complexity onto the user, but also keeps the pool simpler for users that don't need the resizing option. Vault for example has a [`PermitPoolWithMetrics`](https://github.com/hashicorp/vault/blob/5b4b606c0d80820a117cbbda8247b762fceab44b/physical/dynamodb/dynamodb.go#L128-L132) that wraps the permitpool primitive. I think there are other use cases like this that don't necessarily need to be implemented in the pool itself.

Reply:
Ok, then I would suggest that we implement something like a `ResizablePermitPool` that uses a `PermitPool` but provides a `Resize` method. This allows for the mutex to be coupled with the PermitPool, instead of them both being independent package variables. If we want to implement this just in boundary, it wouldn't need be an exported type for now.

Reply:
Sounds good 👍🏻

Reply:
Fixed, please take a look!